setup/Setup.Core/TerrariaExecutableSetter.cs

```diff
@@ -0,0 +1,110 @@
+using Terraria.ModLoader.Setup.Core.Abstractions;
+using Terraria.ModLoader.Setup.Core.Utilities;
+
+namespace Terraria.ModLoader.Setup.Core;
+
+public class TerrariaExecutableSetter
+{
+	private readonly ProgramSettings programSettings;
+	private readonly ITerrariaExecutableSelectionPrompt terrariaExecutableSelectionPrompt;
+	private readonly IUserPrompt userPrompt;
+	private readonly TargetsFilesUpdater targetsFilesUpdater;
+
+	public TerrariaExecutableSetter(
+		ITerrariaExecutableSelectionPrompt terrariaExecutableSelectionPrompt,
+		IUserPrompt userPrompt,
+		TargetsFilesUpdater targetsFilesUpdater,
+		ProgramSettings programSettings)
+	{
+		this.terrariaExecutableSelectionPrompt = terrariaExecutableSelectionPrompt;
+		this.userPrompt = userPrompt;
+		this.targetsFilesUpdater = targetsFilesUpdater;
+		this.programSettings = programSettings;
+	}
+
+	public async Task FindAndSetTerrariaDirectoryIfNecessary(bool noPrompts = false, CancellationToken cancellationToken = default)
+	{
+		if (Directory.Exists(programSettings.TerrariaSteamDir)) {
+			return;
+		}
+
+		if (noPrompts) {
+			Console.WriteLine($"Critical failure. Terraria steam directory '{this.programSettings.TerrariaSteamDir}' does not exist.");
+			Environment.Exit(1);
```

Comment:
Prefer to throw exceptions and let the top level handler print it out.

Reply:
Done